### PR TITLE
Collection Page

### PR DESCRIPTION
**Pull request by filipmanole**
Implementing https://github.com/airswap/airswap-marketplace/issues/22 https://github.com/airswap/airswap-marketplace/issues/11

- Fetching NFT ids by filtering past events.
- Fetching metadata in chunks while scrolling.

### CODE REVIEW

**Comment by filipmanole** [comment]
Short preview:

image

File: src/components/NFTCard/NFTCard.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import './NFTCard.scss';
+
+type NFTCardProps = {
+  imageURI: string;
+  name: string;
+  price: number;
+}
+
+const NFTCard = ({ imageURI, name, price }: NFTCardProps) => (
+  <div className="nft-card">
```
Comment (piersss) [inline src/components/NFTCard/NFTCard.tsx:12]:
```suggestion
  <div className={`nft-card ${className`}>
```

File: src/pages/CollectionPage/CollectionPage.tsx [hunk 2]
```diff
@@ -1,17 +1,72 @@
...
-  <Page
-    className="collection-page"
-    contentClassName="collection-page__content"
-  >
-    <CollectionWidget />
-  </Page>
-);
+const CollectionPage: FC = () => {
+  const { collectionToken } = useAppSelector((state) => state.config);
+  const collectionContract = useContract({ abi: ERC721.abi, address: collectionToken });
```
Comment (piersss) [inline src/pages/CollectionPage/CollectionPage.tsx:16]: contract stuff in the store

File: src/pages/CollectionPage/CollectionPage.tsx [hunk 3]
```diff
@@ -1,17 +1,72 @@
...
+
+  const [tokensData, setTokensData] = React.useState<any[]>([]);
+
+  const index = React.useRef(0);
+  const tokenIds = React.useRef<any[]>([]);
+
+  const fetchCollectionData = async () => {
```
Comment (piersss) [inline src/pages/CollectionPage/CollectionPage.tsx:23]: in the store as a `createAsyncThunk`

File: src/pages/CollectionPage/CollectionPage.tsx [hunk 4]
```diff
@@ -1,17 +1,72 @@
...
+    });
+  }, []);
+
+  return (
+    <Page
+      className="collection-page"
+      contentClassName="collection-page__content"
+    >
+      <CollectionWidget />
+      <div className="nft-list">
```
Comment (piersss) [inline src/pages/CollectionPage/CollectionPage.tsx:59]: nft-list should be in the CollectionWidget

**Review by piersss — CHANGES_REQUESTED** [review]
How to start your PR:

https://github.com/airswap/airswap-marketplace/blob/main/CONTRIBUTING.md#pr-description

File: src/components/NFTCard/NFTCard.tsx [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+
+import './NFTCard.scss';
+
+type NFTCardProps = {
+  imageURI: string;
+  name: string;
+  price: number;
```
[the remark this answers is not in the document]
Comment (filipmanole) [reply src/components/NFTCard/NFTCard.tsx:8]: implemented

**Comment by filipmanole** [comment]
Short video showing the collection page:
https://www.loom.com/share/a023ef4ccc1b4a70a3d405ee20c97bb1

**Review by piersss — APPROVED** [review]
Looks good. Things are working nicely. I added some commits to make it more solid:

1. Your `fetchNFTMetadata` was not setup properly. You only have to make it return a promise and then get the responses in a reducer with the pending, fulfilled, and rejected. So I removed all the collection reducers because they are simply not necessary.

2. You had an `await fetchCallback()` inside the `useInfiniteScroll`. You should keep all async logic inside the store. So I added an `isLoading` to the collection store.

3. I removed `configureCollectionSubscriber`. I understand it's confusing that there's also a `configureBalancesSubscriber` and `configureMetadataSubscriber`. But those subscribers are there because those stores can not be called from a specific `widget`. The collection store is connected to the CollectionWidget, so actions can be called from there.

4. You used a lot of `grid` in your css. It's only necessary when you're using a grid, like the nfts grid. So I left it there only.

5. You had a `onClick` prop on NftCard and used `navigate` for navigation. You should simply use `NavLink` without any listeners.

6. You had classNames like `collection-widget__content__subtitle`, while technically correct it's better to just use `collection-widget__subtitle`.

Things that still need work I have made tickets for:

Infinite scroll and loader needs some work still:
https://github.com/airswap/airswap-marketplace/issues/45

IERC1155 support:
https://github.com/airswap/airswap-marketplace/issues/46

Price of token:
https://github.com/airswap/airswap-marketplace/issues/47

Error handling:
https://github.com/airswap/airswap-marketplace/issues/48

**Pull request merged**